liiatools_pipeline/ops/pnw_census_org.py

```diff
@@ -0,0 +1,149 @@
+from liiatools.common import pipeline as pl
+from liiatools.common.constants import SessionNamesPNWCensusJoins
+from liiatools.common.data import DataContainer
+from liiatools.common.pipeline import open_file
+from liiatools.pnw_census_pipeline.pnw_dataset_join import (
+    join_episode_data,
+    join_header_data,
+    join_uasc_data,
+    join_oc2_data,
+    join_missing_data,
+    join_onspd_data,
+    )
+from liiatools_pipeline.assets.common import workspace_folder, shared_folder
+from liiatools_pipeline.assets.external_dataset import external_data_folder
+
+log = get_dagster_logger(__name__)
+
+
+@op(
+    out={
+        "session_folder": Out(FS),
+    }
+)
+def create_join_session_folder() -> FS:
+    log.info("Creating session folder...")
+    session_folder, session_id = pl.create_session_folder(
+        workspace_folder(), SessionNamesPNWCensusJoins
+    )
+
+    log.info("Opening incoming folder...")
+    session_folder = session_folder.opendir(
+        SessionNamesPNWCensusJoins.INCOMING_FOLDER
+    )
+
+    ssda903_reports_folder = workspace_folder().opendir("current/ssda903/PAN")
+    pl.move_files_for_sharing(
+        ssda903_reports_folder,
+        session_folder,
+        required_table_id=["episodes", "header", "uasc", "oc2", "missing"],
+    )
+
+    pnw_census_reports_folder = workspace_folder().opendir("current/pnw_census/PAN")
+    pl.move_files_for_sharing(pnw_census_reports_folder, session_folder)
+
+    return session_folder
+
+
+@op(
+    ins={
+        "session_folder": In(FS),
+    },
+)
+def joins_pnw_census(
+    session_folder: FS,
+):
+    log.info("Checking necessary files are present...")
+    episodes = re.compile(r"episodes")
+    header = re.compile(r"header")
+    uasc = re.compile(r"uasc")
+    oc2 = re.compile(r"oc2")
+    missing = re.compile(r"missing")
+    pnw = re.compile(r"pnw")
+
+    files = session_folder.listdir("/")
+
+    try:
+        pnw_census_file = next((f for f in files if pnw.search(f)), None)
+    except errors.ResourceNotFound as err:
+        log.error(f"No PNW file to open: {err}")
+        log.info("Exiting run as PNW Census file not available")
+        return
+
+    ssda903_pattern_list = [episodes, header, uasc, oc2, missing]
+
+    # Check which ssda903 files are present
+    ssda903_files = [
+        filename
+        for filename in files
+        if any(pattern.search(filename) for pattern in ssda903_pattern_list)
+    ]
+
+    # Find external data folder for ONSPD file
+    ext_folder = external_data_folder()
+
+    # Open the PNW Census file
+    pnw_census = open_file(session_folder, pnw_census_file)
+    # Derive 'snapshot' date used in every table join equal to the last day of the snapshot month
+    pnw_census["snapshot_date"] = pd.to_datetime(pnw_census[["Year", "Month"]].assign(day=1)) + MonthEnd(0)
+    # Execute logic based on available files
+    if ssda903_files:
+        log.info(f"SSDA903 files found: {ssda903_files}")
+        for matched in ssda903_files:
```

Comment:
In this loop it seems like if one of the files is missing (but not all) it won't raise any error, but the final output will not have that information. I think we should have some sort of notification, then we can decide whether to process at all (if a file is missing) or at least create a blank column for the missing headers (so any linked dashboards still work as expected)

Reply:
Yeah good point